tools/upload-json-dashboard.sh

```diff
@@ -0,0 +1,281 @@
+            ;;
+        j )
+            DASH_JSON=${OPTARG}
+            ;;
+        n )
+            DASH_NAME=${OPTARG}
+            ;;
+        N )
+            DASH_AUTH=0
+            ;;
+        p )
+            DASH_PASSWORD=${OPTARG}
+            ;;
+        P )
+            read -r -p 'Password: ' -s DASH_PASSWORD
+            printf '\n'
+            ;;
+        u )
+            DASH_USERNAME=${OPTARG}
+            ;;
+        v )
+            VERBOSE=$(( VERBOSE + 1 ))
+            ;;
+        x )
+            DASH_EXTRA_CURL=("${OPTARG//[ ]/}")
+            ;;
+        '?' )
+            printf "$ERRFMT" $LINENO "invalid option: -$OPTARG" 1>&2
+            exit 1
+            ;;
+    esac
+done
+shift $((OPTIND -1))
+
+# ================================================================
+# Check required options.
+# ================================================================
+if [ -n "$DASH_JSON" ] && [ -z "$DASH_NAME" ] ; then
+    if jq --version 1>/dev/null 2>/dev/null ; then
+        # If jq is available, extract the variable name.
+        DASH_NAME=$(jq '.__inputs[0].name' "$DASH_JSON" | awk -F'"' '{print $2}')
+    fi
+fi
+# need bash 4 or later
+BASH_MAJOR=$(echo "$BASH_VERSION" | awk -F. '{print $1}')
+if (( BASH_MAJOR < 4 )) ; then
+    # Need 4 or later to support "declare -A".
+    printf "$ERRFMT" $LINENO "this version of bash is too old, must 4 or later: $BASH_VERSION" 1>&2
+    exit 1
+fi
+
+ERRCNT=0
+declare -A ROPTS=( ['-d']="$DASH_DS" ['-j']="$DASH_JSON" ['-g']="$DASH_URL" ['-n']="$DASH_NAME")
+for RKEY in "${!ROPTS[@]}" ; do
+    RVAL=${ROPTS[$RKEY]}
+    if [ -z "$RVAL" ] ; then
+        printf "$ERRFMT" \
+               $LINENO \
+               "missing required option: '$RKEY', see the help (-h) for more information" \
+               1>&2
+        ERRCNT=$(( ERRCNT += 1))
+    fi
+done
+if (( ERRCNT )) ; then exit 1 ; fi
+
+# ================================================================
+# Report the setup.
+# ================================================================
+if (( VERBOSE )) ; then
+    cat <<EOF
+Setup
+    BASH_VERSION    : $BASH_VERSION
+    DASH_AUTH       : $DASH_AUTH
+    DASH_DS         : $DASH_DS
+    DASH_FOLDER     : $DASH_FOLDER
+    DASH_JSON       : $DASH_JSON
+    DASH_NAME       : $DASH_NAME
+    DASH_URL        : $DASH_URL
+    DASH_USERNAME   : $DASH_USERNAME
+    DASH_EXTRA_CURL : ${DASH_EXTRA_CURL[@]}
+    VERBOSE         : $VERBOSE
+EOF
+fi
+
+# ================================================================
+# Other prerequisite checks.
+# ================================================================
+if [ ! -f "$DASH_JSON" ] ; then
+    printf "$ERRFMT" $LINENO "file does not exist: $DASH_JSON" 1>&2
+    exit 1
+fi
+if ! grep '"__inputs"' "$DASH_JSON" 1>/dev/null 2>/dev/null ; then
+    printf "$ERRFMT" $LINENO "missing required record '__inputs' in $DASH_JSON" 1>&2
+    exit 1
+fi
+if ! curl --version --version 1>/dev/null 2>/dev/null ; then
```

Comment:
Did you mean to have two `--version`s? Also, more succinct redirect.
```suggestion
if ! curl --version &>/dev/null ; then
```

Reply:
Yup, that is shorter therefor preferable :-)

Reply:
Looks like you still have the double `--version`.

Reply:
yup, thanks, fixed